### PR TITLE
Remove obsolete compatibility layers from TError

### PR DESCRIPTION
**Pull request by jblomer**
Replaces a few infrastructure bits that have ROOT internal compatibility layers with C++ standard facilities.  Namely:

- `[v]snprintf`
- thread local storage
- varargs

### CODE REVIEW

File: core/base/src/TError.cxx [hunk 1]
```diff
@@ -131,17 +129,12 @@ void ErrorHandler(Int_t level, const char *location, const char *fmt, va_list ap
...
 
    Int_t n = vsnprintf(buf, buf_size, fmt, ap);
-   // old vsnprintf's return -1 if string is truncated new ones return
-   // total number of characters that would have been written
-   if (n == -1 || n >= buf_size) {
-      if (n == -1)
-         buf_size *= 2;
-      else
-         buf_size = n+1;
+   if (n >= buf_size) {
```
Comment (pcanal) [inline core/base/src/TError.cxx:132]: Do you know when (OS and/or compiler) the "old vsnprintf" behavior stopped being implemented?

File: core/base/src/TError.cxx [hunk 2]
```diff
@@ -131,17 +129,12 @@ void ErrorHandler(Int_t level, const char *location, const char *fmt, va_list ap
...
+      buf_size = n+1;
       if (buf != &(small_buf[0])) delete [] buf;
       buf = 0;
       va_end(ap);
-      R__VA_COPY(ap, sap);
+      va_copy(ap, sap);
```
Comment (pcanal) [inline core/base/src/TError.cxx:137]:
```suggestion
      std::va_copy(ap, sap);
```
for consistency (of indicating we are using the c++ version).

File: core/base/src/TError.cxx [hunk 3]
```diff
@@ -196,55 +189,55 @@ void Obsolete(const char *function, const char *asOfVers, const char *removedFro
...
-void Error(const char *location, const char *va_(fmt), ...)
+void Error(const char *location, const char *fmt, ...)
 {
-   va_list ap;
-   va_start(ap,va_(fmt));
-   ErrorHandler(kError, location, va_(fmt), ap);
+   std::va_list ap;
+   va_start(ap, fmt);
+   ErrorHandler(kError, location, fmt, ap);
    va_end(ap);
```
Comment (pcanal) [inline core/base/src/TError.cxx:197]:
```suggestion
   std::va_end(ap);
```

File: core/base/src/TError.cxx [hunk 4]
```diff
@@ -196,55 +189,55 @@ void Obsolete(const char *function, const char *asOfVers, const char *removedFro
...
-void SysError(const char *location, const char *va_(fmt), ...)
+void SysError(const char *location, const char *fmt, ...)
 {
-   va_list ap;
-   va_start(ap, va_(fmt));
-   ErrorHandler(kSysError, location, va_(fmt), ap);
+   std::va_list ap;
+   va_start(ap, fmt);
+   ErrorHandler(kSysError, location, fmt, ap);
    va_end(ap);
```
Comment (pcanal) [inline core/base/src/TError.cxx:208]:
```suggestion
   std::va_end(ap);
```

File: core/base/src/TError.cxx [hunk 5]
```diff
@@ -196,55 +189,55 @@ void Obsolete(const char *function, const char *asOfVers, const char *removedFro
...
-void Break(const char *location, const char *va_(fmt), ...)
+void Break(const char *location, const char *fmt, ...)
 {
-   va_list ap;
-   va_start(ap,va_(fmt));
-   ErrorHandler(kBreak, location, va_(fmt), ap);
+   std::va_list ap;
+   va_start(ap, fmt);
+   ErrorHandler(kBreak, location, fmt, ap);
    va_end(ap);
```
Comment (pcanal) [inline core/base/src/TError.cxx:219]:
```suggestion
   std::va_end(ap);
```

File: core/base/src/TError.cxx [hunk 6]
```diff
@@ -196,55 +189,55 @@ void Obsolete(const char *function, const char *asOfVers, const char *removedFro
...
-void Info(const char *location, const char *va_(fmt), ...)
+void Info(const char *location, const char *fmt, ...)
 {
-   va_list ap;
-   va_start(ap,va_(fmt));
-   ErrorHandler(kInfo, location, va_(fmt), ap);
+   std::va_list ap;
+   va_start(ap, fmt);
+   ErrorHandler(kInfo, location, fmt, ap);
    va_end(ap);
```
Comment (pcanal) [inline core/base/src/TError.cxx:230]:
```suggestion
   std::va_end(ap);
```

File: core/base/src/TError.cxx [hunk 7]
```diff
@@ -196,55 +189,55 @@ void Obsolete(const char *function, const char *asOfVers, const char *removedFro
...
-void Warning(const char *location, const char *va_(fmt), ...)
+void Warning(const char *location, const char *fmt, ...)
 {
-   va_list ap;
-   va_start(ap,va_(fmt));
-   ErrorHandler(kWarning, location, va_(fmt), ap);
+   std::va_list ap;
+   va_start(ap, fmt);
+   ErrorHandler(kWarning, location, fmt, ap);
    va_end(ap);
```
Comment (pcanal) [inline core/base/src/TError.cxx:241]:
```suggestion
   std::va_end(ap);
```

File: core/base/src/TError.cxx [hunk 8]
```diff
@@ -254,10 +247,10 @@ void Warning(const char *location, const char *va_(fmt), ...)
...
-void Fatal(const char *location, const char *va_(fmt), ...)
+void Fatal(const char *location, const char *fmt, ...)
 {
-   va_list ap;
-   va_start(ap,va_(fmt));
-   ErrorHandler(kFatal, location, va_(fmt), ap);
+   std::va_list ap;
+   va_start(ap, fmt);
+   ErrorHandler(kFatal, location, fmt, ap);
    va_end(ap);
```
Comment (pcanal) [inline core/base/src/TError.cxx:255]:
```suggestion
   std::va_end(ap);
```

File: core/base/src/TError.cxx (hunk 1 above)
Comment (jblomer) [reply core/base/src/TError.cxx:132]: Some googling reveals the following:

> Some of the first implementations of sprintf() (before the
> C99 standard, e.g. the GNU libc 5) did return -1 if the
> resulting string did not fit into memory allowed for writing.

Apparently the same for old MS compilers, e.g. VC++ 6.0.  

In any case, the new behavior was standardized with C99 and C++11, so I guess we should be good.

File: core/base/src/TError.cxx (hunk 2 above)
Comment (jblomer) [reply core/base/src/TError.cxx:137]: Unless I'm mistaken, `cstdio` does not put the functions/macros `va_copy` nor `va_end` in the `std` namespace.  Only the data type `va_list`.

File: core/base/src/TError.cxx (hunk 1 above)
Comment (pcanal) [reply core/base/src/TError.cxx:126]: Make sure this info is in the commit message.

File: core/base/src/TError.cxx (hunk 2 above)
Comment (pcanal) [reply core/base/src/TError.cxx]: humm .. right :( ... they are still C macros  ....

**Pull request merged**